src/celeritas/ext/detail/GeantSurfacePhysicsLoader.cc

```diff
@@ -0,0 +1,426 @@
+/*!
+ * Construct with \c SurfacePhysics input to be filled by \c operator() .
+ */
+GeantSurfacePhysicsLoader::GeantSurfacePhysicsLoader(inp::SurfacePhysics& result)
+    : result_(result)
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Populate surface physics data for a given \c SurfaceId .
+ */
+void GeantSurfacePhysicsLoader::operator()(SurfaceId sid)
+{
+    CELER_EXPECT(sid);
+    detail::GeantSurfacePhysicsHelper helper(sid);
+    try
+    {
+        this->insert_reflectivity(helper);
+        this->insert_roughness(helper);
+        this->insert_interaction(helper);
+        this->validate_model(helper);  // Verify model requirements
+    }
+    catch (RuntimeError const& e)
+    {
+        throw;
+    }
+
+    CELER_LOG(debug) << "Inserted surface id " << sid.unchecked_get()
+                     << " with " << to_cstring(helper.surface().GetModel())
+                     << " model and "
+                     << to_cstring(helper.surface().GetFinish()) << " finish";
+}
+
+//---------------------------------------------------------------------------//
+// PRIVATE MEMBER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Collect reflectivity information from a given optical surface.
+ */
+void GeantSurfacePhysicsLoader::insert_reflectivity(
+    detail::GeantSurfacePhysicsHelper& helper)
+{
+    auto sid = helper.surface_id();
+    inp::ReflectivityModels refl_mods;
+    if (!this->analytic_reflection_only(helper.surface()))
+    {
+        // Insert any model that includes user-defined grid reflectivity
+        inp::ReflectionGrid refl_grid;
+        helper.get_property(&refl_grid.grid, "REFLECTIVITY");
+        refl_mods.grid.insert({sid, std::move(refl_grid)});
+    }
+    refl_mods.analytic.insert({sid, inp::ReflectionAnalytic{}});
+    result_.reflectivity = std::move(refl_mods);
+    CELER_ENSURE(result_.reflectivity);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Collect roughness information from a given optical surface.
+ */
+void GeantSurfacePhysicsLoader::insert_roughness(
+    detail::GeantSurfacePhysicsHelper& helper)
+{
+    using G4OSM = G4OpticalSurfaceModel;
+    using G4OSF = G4OpticalSurfaceFinish;
+
+    auto sid = helper.surface_id();
+    auto const& surf = helper.surface();
+    auto const g4model = surf.GetModel();
+    switch (g4model)
+    {
+        case G4OSM::glisur: {
+            if (surf.GetFinish() == G4OSF::polished)
+            {
+                // Perfectly polished surface
+                result_.roughness.polished.insert({sid, inp::Polished{}});
+            }
+            else
+            {
+                // Smearing is available (surf.GetFinish() == G4OSF::ground)
+                // Celeritas' roughness is the complement of Geant4 polish
+                inp::SmearRoughness smear{real_type{1} - surf.GetPolish()};
```

Comment:
Is  a polish of 1 used in practice? Should we replace it with a polished model?

Reply:
Actually, re-re-reading the code, having `ground` finish and `polish = 1.0` is still possible. So you can have a perfectly polished surface that undergoes spike reflection (with finish == polished), or a ground finish that is also perfectly polished (with `GetPolish()` returning 1), but undergoing lobe reflection...

Reply:
Aren't spike vs lobe the reflection model, not the surface facet distribution?